### ISSUE TITLE
TimeSpan SqlLite

### ISSUE THREAD

**Issue by xmedeko** [StackExchange/dapper-dot-net#465]
I use `Dapper.Contrib` with SqlLite and it does not work.
```sql
create table mytable (
    id integer primary key,
    duration text not null,
);
insert into mytable(duration) values ('0.01:00:00');
```
C# code
```c#
[Table("mytable")]
public class ClientTest
{
        public long Id { get; set; }
        public TimeSpan Duration { get; set; }
}
```
Got error:
```
System.Data.DataException: Error parsing column 4 (duration=0.01:00:00 - String)
```

When I write a `TypeHandler`:
```c#
public class TimeSpanHandler : SqlMapper.TypeHandler<TimeSpan>
{
    public override TimeSpan Parse(object value)
    {
        return TimeSpan.Parse((string)value);
    }

    public override void SetValue(IDbDataParameter parameter, TimeSpan value)
    {
        throw new InvalidOperationException("Never called!");
        //parameter.Value = value.ToString();
    }
}
```
Then it works fine for query and update, too. That's a mistake since `SetValue` is never called.

So we have a 2 bugs:

1. `TimeSpan` for SqlLite does not work by default.
2. TypeHandler<TimeSpan> is called just `Parse`, the `SetValue` is never called.

**Comment by mgravell** [comment]
SetValue is used for parameters. I don't see that in your example - am I
right in this?

I agree it would be nice if the qlite provider worked nicely here. What I
could do is check whether the typed data reading API works - but I know
there are providers that don't support some parts of that API. Which is
very problematic, from the perspective of a general purpose library!

**Comment by NickCraver** [comment]
@mgravell I had this in linqpad, thinking maybe we should make the int/long to/from conversion (assuming ticks) an in-box conversion - the long constructor for TimeSpan (seems to be common practice elsewhere). Which typed data API are you talking about for TimeSpan?

**Comment by mgravell** [comment]
You're right, @NickCraver: there no typed API for TimeSpan. My mistake. A
few others, but not that.

Agree that this should be stored as int/long (for sortability); that is
doable in the read loop, but I can't think of any sane way to do that
during parameter creation.

But having it work for read would be a good start :)

**Comment by NickCraver** [comment]
I have this working here locally - did you not register the type handler perhaps?

Working code:

```C#
void Main()
{
	SqlMapper.AddTypeHandler(typeof(TimeSpan), new TimeSpanHandler());
	using (var conn = new SQLiteConnection("Data Source=:memory:;"))
	{
		var result = conn.QueryFirstOrDefault<ClientTest>(@"
create table mytable (
    id integer primary key,
    duration text not null
);
insert into mytable(duration) values ('0.01:00:00');

Select * From mytable");
		Console.WriteLine($"Hours: {result.Duration.Hours}");
	}
}

public class ClientTest
{
	public long Id { get; set; }
	public TimeSpan Duration { get; set; }
}

public class TimeSpanHandler : SqlMapper.TypeHandler<TimeSpan>
{
	public override TimeSpan Parse(object value)
	{
		return TimeSpan.Parse((string)value);
	}

	public override void SetValue(IDbDataParameter parameter, TimeSpan value)
	{
		throw new InvalidOperationException("Never called!");
		//parameter.Value = value.ToString();
	}
}
```

**Comment by NickCraver** [comment]
Ah okay - wasn't sure which set of methods you were thinking of and maybe I wasn't - sanity! `TimeSpan` isn't `IConvertible` so yeah, parameters will be fun. Kid time here - I'll poke a bit later...many issues to follow up on tonight.

**Comment by mgravell** [comment]
I believe the point of the bug report is that a custom handler is required
*at all*, rather than dapper just making it work.

**Comment by NickCraver** [comment]
Wow, I can't read at all - yeah you're right. I suck. Okay, let's look at making this work with `long`s across all providers. Maybe *not* `int` so people are less likely to blow off lower extremities that way.

**Comment by mgravell** [comment]
I already special-case at least one type (I forget which). I could special
case TimeSpan and support int, long and string (using invariant). That
would be fairly minimal.

**Comment by mgravell** [comment]
Of course, `int` and `long` still have the issue of scale. What is `10`? ticks? millis? seconds? ice-ages?

**Comment by NickCraver** [comment]
That's why I'd lobby for `long` and `string` only, as `.Ticks` or `.ToString()` or `TimeSpan`. It's the highest fidelity available, so that should be the storage mechanism IMO (this seems fairly standard from a bit of research).

**Comment by mgravell** [comment]
like that?

**Comment by NickCraver** [comment]
+1 - LGTM

**Comment by xmedeko** [comment]
This bug has two points: one is `TimeSpan` default deserialization, I think just `string` (mgravell's fix) is fine.

The other bug is that a `TypeHandler` for such "internal" type is called only for deserialization - `Parse()` and not for serialization - `SetValue()`. (At least for Dapper.Contrib). It should be either both (I prefer) or none of them. When I register `TypeHandler` for my class, then both `Parse()` and `SetValue()` are called correctly.

_Remark:_ As a workaround now I've made a class `TimeSpanWrapper` with implicit cast to `TimeSpan`, registered the TypeHandler for it and it works ok.

**Comment by mgravell** [comment]
Re the "other bug" - the point I was making there is that the example code
shown *would not be expected* to use SetValue, because it never passes such
in as a parameter.

Note there's an important point here, though: the two "bugs" reported are
kinda mutually exclusive. If we make the type-handler work fully in this
case, it makes it moot to parse automatically. Note that what we *can't* do
is make it parse *and* write automatically, because in the general case:
the ADO.NET provider might work fine with TimeSpan being passed in, or we
can't second-guess the non-intrinsic storage type chosen by the user; as
demonstrated by your decision to use string in this case when many others
have used long.

Btw: to reiterate, using a string as the storage for a timespan is ...
dubious. It isn't sortable, etc, and requires lots of extra work. If that
is what your existing schema demands... well, I guess that's what it is. In
which case, a type-handler seems the right approach. If the right "fix"
here is to make type-handlers work fully, we can do that. It isn't a unique
report. I was very reluctant to add it in the case of things like "int"
(I'm struggling to think of a sane ADO.NET provider that can't handle int),
but it makes a lot more sense in the case of things like TimeSpan.

On 22 February 2016 at 07:56, Ondrej Medek <notifications@github.com> wrote:

> This bug has two points: one is TimeSpan default deserialization, I think
> just string (mgravell's fix) is fine.
>
> The other bug is that a TypeHandler for such "internal" type is called
> only for deserialization - Parse() and not for serialization - SetValue().
> (At least for Dapper.Contrib). It should be either both (I prefer) or none
> of them. When I register TypeHandler for my class, then both Parse() and
> SetValue() are called correctly.
>
> *Remark:* As a workaround now I've made a class TimeSpanWrapper with
> implicit cast to TimeSpan, registered the TypeHandler for it and it works
> ok.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/StackExchange/dapper-dot-net/issues/465#issuecomment-187060156>
> .
>



-- 
Regards,

Marc

**Comment by xmedeko** [comment]
Yes, the example is incomplete, **neither** the code for query or update is included. I use 'Dapper.Contrib':  `Get<MyTable>(id)` and `Update(conn, object)`. When I create `TimeSpanWrapper` and register `TypeHandler<TimeSpanWrapper>` for it, then both `Parse()` and `SetValue()` are called, so I can make any conversion I like.

The problem is in **inconsistent** Dapper behaviour. The `SetValue()` is called for `TypeHandler<TimeSpanWrapper>`, but not for `TypeHandler<TimeSpan>`. The `Parse()` is called always.

I agree making `TypeHandler<TimeSpan>` fully working is more important, so the user can choose whatever serialization she likes. Anyway, to have working default serialization is always nice. I also do not know if the problem is in the Dapper or `System.Data.SQLLite` provider. It the bug is on the SQLLite side, we should report it to the SQLite team and mayby Dapper may leave it as is.

I do now vote for `string` storage and I do not use it. I just think any solution working out of the box is fine, even `string`. Note, the SQLite uses string for `DateTime` storage by default, although it can do `DateTime.Ticks`, too. They have this piece of code with remark, which I do not understand:
```c#
public enum SQLiteDateFormats
{
    /// <summary>
    /// Use the value of DateTime.Ticks.  This value is not recommended and is not well supported with LINQ.
    /// </summary>
    Ticks = 0,
/// ...
}
```

Summary: I do not know where's the problem with `TimeSpan` default serialization (Dapper or SQLLite). I do not vote for `string` or `long` as default conversion. IMHO if Dapper (Dapper.Contrib) makes the `TypeHandler<TimeSpan>` fully working I am happy since I do not need the `TimeSpanWrapper` any more.

**Comment by xmedeko** [comment]
Note also #319.

**Comment by NickCraver** [comment]
Tagging this for v3 for #688 - if we move the types to `IDapper` and such, we could be per-provider much more sanely (with people opt-ing in or something) and this is worth considering in our design approach to if it makes sense.
